### ISSUE TITLE
Invalid type inference when calling function with overloads

### ISSUE THREAD

**Issue by giuseppelt** [microsoft/TypeScript#50696]
# Bug Report
Some strange type checking/inference behavior with function overloads.
Compiler reports error `'resource' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer`.

See on playground link:
- Bug1: inference
If the line 32 is active, the line 15 will report error (`resource` variable is not inferred).
If the line 32 is commented, no error is reported (`resource` is inferred ok).

- Bug2: overloads
With the line 32 active
If the `_raiseError` overloads are commented (line 46 and 47), no error is reported.
With the overloads, the typing error is reported.

- Bug3: nullable
With the line 32 active
if the `getAccountPlan` on line 39 returns `Promise<UsagePlan>` no error is reported,
if it returns `Promise<UsagePlan | undefined>`, the error is reported

### 🔎 Search Terms
typing inference, control flow analysis

### 🕗 Version & Regression Information
v4.8, I didn't try on previous versions

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwDYEMDOa4FU0oObADKwUAbgJYLBwDeAUHXE3OgJ4B2Cc5nUw6wABQoECCAFd2MAJIATAFxw0MKDzwAaOHzQSoVNIoBKwMVFkAeZavYa47cQFsARiQB8ASlqNmPse2VwwFKqwJgAvHAA8k4AViYwAHRBKuShgtq6+u4A3N4+TH4BYKjscBEoAO4o5PAwABbkaAkEMACCohJSAAolwh2SMrI5efnkAGZwggCExSjsnvT5S1rAMOJQ7LnLcAC+DNuF8OK4BOEsVTVw9Y3Nqzj4wABqKEjiaSJiA3Ka0XEIiQBrYCsNDpUKZULuYYjHxjaCTQ5wADaQNYmhQDk6MAAunAIBNkiE0AsYctxpMMViymEIgAGTx+GA8N65UlLREZdZUMpwWbsBKcvShBJjHiyQQgMquOAgBKo6kRVHDbY+cnTQVUEkq5aM5nALba3b7Q1wqAIiD+eBIcgOS74lY6LnABLW20wYleQ2jCaCV01BIwVhgahTGlwABE9W0dQgSFk4a1XuWUYgFTswDTAFEoFBoIJwwmDUmjWyVYjjg8eRXTiKxRKpTKBeCnQB9eU0iIa53ygBkPcbwdUEFkCrgfsSg-Iw+hxYKFoCpBebyrJ2AAH4EovXtQAD47uC0uAAahYmIGRaTaq3y+l44DdWjsaGntnPmuTRbUGqaGA2dzUHzCBSBIFso1CGM43DTRFlfZY+UUPkEnIWR1FLV8u0UVFUNguCSCnBQxxtf1J2HTQAHoyLgcwAFpqMQOYAHIjh-Qi3TgOoSGATRA2DNAEFUMB4CccQ8DQw0dmVWc9i9aTllkpg9hGNhODgFp2k+bpeg+LE5EUKw1HcRQulzW0f3Me4CB6OY4H3SRZGAUV2GAWRpRgpY+DWDY4DshyeGcqYRnklg0A4LgWgsp4l3efopF0pQUhsTQuzQNc9ISvAkWxQy4GMiBTOAczV0zYJWEy1zSw89ZSh8xz-IvRSfDAXMYHiZy4E-b9fxzPMSH-NLrFsWQUBgFBUrgYxTAsfTEpYdhWA8RQnOAqALyaiAWv+NqOsaLr-0EXroH6tRNAcUJVzG6bBuG0ajBMaApvS9E5oW9NltW5rWpHbafz-HruqgI6ZtOjAHgu9KbPGu6zEsR7ZvmzQhpGsaJvumGBqe+bsqWkgXyWFM0ycrN-v2-7JIUuhFLoHjqAi4qVFYHk3O4AjLovbSBhbZDAbwVa8OHRQIq6PnZAvLs22BbmL2vYBFscFwVpGPgxk4zhgBbRGZbgAARYbdzscQkCQC9+P4FrZHV3XFB1lqL3EMANfNjWrd13JFNASBYCuIMadXIWhxHCJwy11oAE1wwh8MAFlIgAOQAFQACXDBh3egWpvewX2SkZkYufigaRjEezucV5shQMTOHmMR0hS13z2BqKd2EyimU-ANOveDSuCGriE68cxuLRznxURLnx2Axdcx+Yey+IEpkLTBgvVT8Je1BGccK4igAZIiYH7nhB+b7FW7oVPPep7vgF3t0D4bhfSjCEZ9wiuOH3Ap8b5qNuPfTrvX-fmgCCsgv7wAiEzamihIyAOAcnRqwsBa+2FiMMCQCnyy2cCQVuQA)

### 💻 Code

```ts
for (const [key, amount] of entries) {
    const resource = plan.resources.find(x => x.key === key); // <--- typescript will complain here if the line with limit.period is used
    if (!resource) continue;

    for (const limit of resource.limits) {
        if (limit.type !== "threshold") throw new Error("");

        const usage = usages.find(x => x.resource_key === resource.key && x.period === limit.period);
        const value = usage?.value || 0 + amount;
        if (value > limit.threshold) {
            this._raiseError("over_threshold", {
                plan: plan.id,
                resource: key,
                period: limit.period, // <-- can't use limit here, typescript will report error above, but it works if commented out
            });
        }
    }
}
```

### 🙁 Actual behavior
Type error

### 🙂 Expected behavior
No error reported

**Comment by jcalz** [comment]
(I'm not a TS team member)

Duplicate of or strongly related to #35638.   

> I didn't try on previous versions

The playground lets you check this sort of thing easily.  Looks like this goes back to TS3.7.


---

The type checker can hit circular type dependencies when it's inferring types, even though a human being might analyze the types in a different way and not hit such a circularity.  In this case it looks like an interaction between overload resolution and reachability / control flow analysis that makes the compiler give up.  You could write a manual annotation on almost any of the things that are complaining to make the error go away, if you want to work around it (which you probably should).

Here's a more minimal repro:

```ts
interface UsagePlan {
    resources: { limits: { period: number }[]; }[]
}

export class UsageService {
    increase() {
        const plan = null! as UsagePlan | undefined;
        if (!plan) { return; }
        for (const _ of []) {
            const resource = plan.resources[0];
            for (const limit of resource.limits) {
                this._raiseError({ period: limit.period });
            }
        }
    }
    _raiseError(data?: Record<string, any>): never;
    _raiseError(message?: string, data?: Record<string, any>): never;
    _raiseError() { throw new Error() }
}
```
[Playground link](https://www.typescriptlang.org/play?ts=4.8.2#code/JYOwLgpgTgZghgYwgAgKoGc4HMIAUA2cIyA3gFDKXJQToD2ArlEugFynL7AC2wYbHAA7RgdACbsQDbgCNoyAL4BtALoBuRarIKyZCAA9BdKGGQJC6dGkw4AytABuwJKQpVQCGnHQQAFAEpXKmCzOhB0U0FCYgBeZCl8fABCZG9rbDxo5AAfZAYQMQgYUAgxNTcQ5GAYZF8kqKJAkmoIMCYQDR1KyhhjWoQwiOQAfWQ6GtUmiu7KAfDTGnomFziGkAA6RcZmWiUABnVpmd6ofsHTLl5TcZalnfXLvnQpmZmwAAtgdHXhqDgviAAUSgUGMvmawigogknB4fHWkOhin85VeVC63QxlCxIz+AOBoKgvjEcDAcAA-OwAEoQAZQMQAHgiUJAWAANKkQABPAB8-kkEAc0FRVF+-x8BLB3FoNgglOQzNA7OQJLJ8ppdMZitZHKIvP58UFwumYvxILBTWQH1BAHdDXbJUTAjodGQgA)

Looks like the type checker has to figure out:
* which call signature `this._raiseError({period: limit.period})` uses, which depends on
* the type of `{period: limit.period}`, which depends on
* the type of `limit`, which depends on
* the type of `resource`, which depends on
* the type of `plan`, which depends on
* control flow analysis... `plan` is narrowed by a falsiness check before the loop, and in the loop its type depends on whether or not it's been reassigned or guarded in other ways, which depends on
* whether `this._raiseError()` affects reachability by returning `never`, (probably why this breaks [starting with TS3.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#better-support-for-never-returning-functions)) which depends on
* which call signature  `this._raiseError({period: limit.period})` uses

And there's a circularity.  A human being could easily analyze her way past almost any of these; surely the first call signature is being chosen because it's not a string; surely `plan` can only be `UsagePlan` or `undefined` and `undefined` wouldn't even let you get to the lines we're analyzing; surely `plan` is not being reassigned in the loop; surely `this._raiseError()` returns `never` no matter which call signature is being called; etc.  But we don't have human beings, we have a compiler.   The question here is whether or not there's a *bug* in the compiler, and it doesn't look like it.
